File: 3.x.x/crates-utopia/emails/src/canonicals/gmail.rs
use crate::canonicals::provider::{
    remove_dots, remove_plus_addressing, require_non_empty_local, to_lower_case, Canonical,
    Provider,
};
use crate::EmailError;

/// Gmail / Googlemail canonicalization (PHP `Gmail`).
#[derive(Debug, Clone, Copy, Default)]
pub struct Gmail;

const SUPPORTED_DOMAINS: &[&str] = &["gmail.com", "googlemail.com"];
const CANONICAL_DOMAIN: &str = "gmail.com";

impl Provider for Gmail {
    fn supports(&self, domain: &str) -> bool {
        SUPPORTED_DOMAINS.contains(&domain)
    }

    fn get_canonical(&self, local: &str, _domain: &str) -> Result<Canonical, EmailError> {
        let normalized = to_lower_case(local);
        let normalized = remove_plus_addressing(&normalized);
        let normalized = remove_dots(&normalized);
        let normalized = require_non_empty_local(normalized)?;
        Ok(Canonical {
            local: normalized,
            domain: CANONICAL_DOMAIN.to_string(),
        })
    }

    fn get_canonical_domain(&self) -> &'static str {
        CANONICAL_DOMAIN
    }

    fn get_supported_domains(&self) -> &'static [&'static str] {
        SUPPORTED_DOMAINS
    }
}

File: packages/emails/src/Canonicals/Providers/Gmail.php
<?php

namespace Utopia\Emails\Canonicals\Providers;

use Utopia\Emails\Canonicals\Provider;

/**
 * Gmail
 *
 * Handles Gmail and Googlemail email normalization based on validator.js rules
 * - Removes all dots from local part
 * - Removes plus addressing (subaddress)
 * - Normalizes to gmail.com domain
 * - Converts googlemail.com to gmail.com
 */
class Gmail extends Provider
{
    private const SUPPORTED_DOMAINS = ['gmail.com', 'googlemail.com'];

    private const CANONICAL_DOMAIN = 'gmail.com';

    public function supports(string $domain): bool
    {
        return in_array($domain, self::SUPPORTED_DOMAINS, true);
    }

    public function getCanonical(string $local, string $domain): array
    {
        // Convert to lowercase
        $normalizedLocal = $this->toLowerCase($local);

        // Remove plus addressing (subaddress) - everything after +
        $normalizedLocal = $this->removePlusAddressing($normalizedLocal);

        // Remove dots from local part
        $normalizedLocal = $this->removeDots($normalizedLocal);

        // Ensure local part is not empty after normalization
        if (empty($normalizedLocal)) {
            throw new \InvalidArgumentException('Email local part cannot be empty after normalization');
        }

        return [
            'local' => $normalizedLocal,
            'domain' => self::CANONICAL_DOMAIN,
        ];
    }

    public function getCanonicalDomain(): string
    {
        return self::CANONICAL_DOMAIN;
    }

    public function getSupportedDomains(): array
    {
        return self::SUPPORTED_DOMAINS;
    }
}
